hs-bindgen/src/HsBindgen/Hs/AST.hs

```diff
@@ -0,0 +1,110 @@
+-- | Haskell AST
+--
+-- This is a bespoke AST, specifically targetted at the hs-bindgen use case. We
+-- do not attempt to cover all of Haskell, and we provide a more strongly typed
```

Comment:
I'm not sure I like this.

From the backend perspective it would be nicer to have as small AST as possible to work with, not dozens of possible case splits, instance declarations etc. But from constructing, frontend perspective I understand that it's nicer to have more "sugar".

I'd argue that there should be concrete low-level AST. Whether the high-level is also data-types with conversion functions to low-level AST, or directly functions to construct low-level AST, I don't have good opinion about, both have own merits.

Reply:
An argument for having low-level (close to syntax) haskell AST is that in perfect case, when adding new frontend functionality we wouldn't need to extend all backends (we only have two, but still). Now touching both backends will be in most cases necessary.

Reply:
Yeah, changing some of these to be shallow rather than deep embeddings might be a better option, that's certainly something we can tweak.

Reply:
Will simplify this.

Reply:
I've thought about this some more. We currently have two kinds of specific "kinds" of concepts in the AST:

* `StorableInstance`. We could replace this with a general `Instance` concept.
* `PeekByteOff`, `PokeByteOff`. We could replace this with general `App` expression.

In both of these cases, we have the problem that we'd still like to refer to specific concepts in the `base` libraries, such as `Foreign.Storable.sizeOf`. We could represent these simply as strings, and then match against those strings in the TH backend to resolve to `'Foreign.Storable.sizeOf'`, but this feels quite unpleasant. We could instead TH-side use `mkName` and rely on imports, but that doesn't feel quite right either? 

We could resolve this by introducing something like `data Name = Generated String | Global GlobalName`, where `Generated` would correspond to TH `mkName` (for types we generate during generation), and `data GlobalName = ForeignStorableSizeOf | ForeignStorableAlignment | ...`, where we could then resolve the latter probably.

I was halfway implementing this, but then stopped again. There is another benefit to the AST representation as it is now: it can only represented Well-Typed Haskell programs! This is something that I think is actually quite a useful benefit. I don't to worry that the code I generate in the AST might be ill-typed, only to realize this when I try it out. (Provided no bugs in the TH generation, of course.) 

Since we will generate AST values probably in more places than we modify the backend, this feels like a useful benefit at not too much cost? So let's keep it for now (but still changing the `Vec` thing). We can always reconsider again if it turns out the benefit isn't worth the extra cost in the backends.

Reply:
Sorry for going back and forth on this, but of course we could still achieve the type safety through the high-level API. Let me actually write some more code, experiment a bit more, and then come to a conclusion.

Reply:
> it can only represented well-typed Haskell programs! 

Surely it can if it cannot even enforce arity of functions! Missing instances for `peekByteOff` etc. It's quite a loose guarante.

But I guess we'll just make enough helpers function to generate case binds etc. from each specific construct  (as we already have helpers like `simpleMatch`).

Reply:
True :)

By the end of today I hope to have a version which reduces duplication between the two backends as well as uses vectors, which I hope you will be happier with :)